Source/WebKit/UIProcess/ios/WKContentView.mm

```diff
@@ -921,38 +926,72 @@ - (NSUInteger)_wk_pageCountForPrintFormatter:(_WKWebViewPrintFormatter *)printFo
             pageCount = _page->computePagesForPrintingiOS(frameID, printInfo);
     }
 
-    ASSERT(!_pdfPrintCompletionSemaphore);
-    if (isPrintingOnBackgroundThread)
-        _pdfPrintCompletionSemaphore = makeUnique<BinarySemaphore>();
-
-    // Begin generating the PDF in expectation of a (eventual) request for the drawn data.
-    _pdfPrintCallbackID = _page->drawToPDFiOS(frameID, printInfo, pageCount, [retainedSelf = retainPtr(self)](RefPtr<WebCore::SharedBuffer>&& pdfData) {
-        if (!pdfData || pdfData->isEmpty())
-            retainedSelf->_printedDocument = nullptr;
-        else {
-            auto data = pdfData->createCFData();
-            auto dataProvider = adoptCF(CGDataProviderCreateWithCFData(data.get()));
-            retainedSelf->_printedDocument = adoptCF(CGPDFDocumentCreateWithProvider(dataProvider.get()));
-        }
+    {
+        Locker locker { _pdfPrintCallbackMutex };
 
-        if (auto pdfPrintCompletionSemaphore = std::exchange(retainedSelf->_pdfPrintCompletionSemaphore, nullptr))
-            pdfPrintCompletionSemaphore->signal();
+        ASSERT(!_pdfPrintCompletionCondition);
+        if (isPrintingOnBackgroundThread)
+            _pdfPrintCompletionCondition = makeUnique<Condition>();
 
-        retainedSelf->_pdfPrintCallbackID = 0;
+        // Mark the callback as pending.
+        _pdfPrintCallbackID = 0;
+    }
+
+    ensureOnMainRunLoop([frameID, printInfo, pageCount, self, retainedSelf = retainPtr(self)] {
+        // Begin generating the PDF in expectation of a (eventual) request for the drawn data.
+        auto callbackID = _page->drawToPDFiOS(frameID, printInfo, pageCount, [retainedSelf = retainPtr(retainedSelf.get())](RefPtr<WebCore::SharedBuffer>&& pdfData) {
+            if (!pdfData || pdfData->isEmpty())
+                retainedSelf->_printedDocument = nullptr;
+            else {
+                auto data = pdfData->createCFData();
+                auto dataProvider = adoptCF(CGDataProviderCreateWithCFData(data.get()));
+                retainedSelf->_printedDocument = adoptCF(CGPDFDocumentCreateWithProvider(dataProvider.get()));
+            }
+
+            Locker locker { retainedSelf->_pdfPrintCallbackMutex };
+
+            if (auto pdfPrintCompletionCondition = std::exchange(retainedSelf->_pdfPrintCompletionCondition, nullptr))
```

Comment:
This code (nulling and notifying the condition + resetting the callbackID) is duplicated above so we may want to factor it out to its own function.

Reply:
Done.